consensus/src/quorum_store/utils.rs

```diff
@@ -188,159 +215,162 @@ impl ProofQueue {
-                    if *expiration < current_block_timestamp {
-                        counters::GAP_BETWEEN_BATCH_EXPIRATION_AND_CURRENT_TIME_WHEN_PULL_PROOFS
-                            .observe((current_block_timestamp - expiration) as f64);
+                        break;
                     }
                 }
+                if num_remaining != num_read {
+                    author_to_num_remaining.insert(peer, num_remaining - num_read);
+                }
             }
-            size -= 1;
         }
         info!(
             // before non full check
             byte_size = cur_bytes,
             block_size = cur_txns,
             batch_count = ret.len(),
-            remaining_proof_num = size,
-            initial_remaining_proof_num = initial_size,
             full = full,
             return_non_full = return_non_full,
             "Pull payloads from QuorumStore: internal"
         );
 
         if full || return_non_full {
-            counters::EXPIRED_PROOFS_WHEN_PULL.observe(num_expired_but_not_committed as f64);
             counters::BLOCK_SIZE_WHEN_PULL.observe(cur_txns as f64);
             counters::BLOCK_BYTES_WHEN_PULL.observe(cur_bytes as f64);
             counters::PROOF_SIZE_WHEN_PULL.observe(ret.len() as f64);
+            counters::EXCLUDED_TXNS_WHEN_PULL.observe(excluded_txns as f64);
             ret
         } else {
             Vec::new()
         }
     }
 
-    pub(crate) fn num_total_txns_and_proofs(&mut self, current_block_timestamp: u64) -> (u64, u64) {
-        let mut remaining_txns = 0;
-        let mut remaining_proofs = 0;
-        // TODO: if the digest_queue is large, this may be too inefficient
-        for (digest, expiration) in self.digest_queue.iter() {
-            // Not expired
-            if *expiration >= current_block_timestamp {
-                // Not committed
-                if let Some(Some(proof)) = self.digest_proof.get(digest) {
-                    remaining_txns += proof.num_txns();
-                    remaining_proofs += 1;
+    pub(crate) fn handle_updated_block_timestamp(&mut self, block_timestamp: u64) {
+        assert!(
+            self.latest_block_timestamp <= block_timestamp,
+            "Decreasing block timestamp"
+        );
+        self.latest_block_timestamp = block_timestamp;
+
+        let peers: Vec<_> = self.author_to_batches.keys().cloned().collect();
+        let mut num_expired_but_not_committed = 0;
+
+        for peer in peers {
+            let mut queue = self.author_to_batches.remove(&peer).unwrap();
+            let num_expired = queue
+                .iter()
+                .take_while(|batch| batch.expiration() < block_timestamp)
+                .count();
+
+            for batch in queue.drain(0..num_expired) {
+                if self
+                    .batch_to_proof
+                    .get(&batch)
+                    .expect("Entry for unexpired batch must exist")
+                    .is_some()
```

Comment:
Following up on the comment below, why do we need unexpired, committed batch to exist?